realm/realm-library/src/main/java/io/realm/ManagedMapManager.java

```diff
@@ -209,38 +233,145 @@ public Mixed put(Object key, Mixed value) {
  */
-class BooleanValueOperator extends MapValueOperator<Boolean> {
+class BoxableValueOperator<T> extends MapValueOperator<T> {
 
-    BooleanValueOperator(BaseRealm baseRealm, OsMap osMap) {
-        super(baseRealm, osMap);
+    BoxableValueOperator(BaseRealm baseRealm, OsMap osMap, ClassContainer classContainer) {
+        super(baseRealm, osMap, classContainer);
     }
 
     @Override
-    public Boolean get(Object key) {
-        return (Boolean) osMap.get(key);
+    public T get(Object key) {
+        //noinspection unchecked
+        return (T) osMap.get(key);
     }
 
     @Override
-    public Boolean put(Object key, Boolean value) {
-        Boolean original = (Boolean) osMap.get(key);
+    public T put(Object key, T value) {
+        //noinspection unchecked
+        T original = get(key);
         osMap.put(key, value);
         return original;
     }
+}
+
+/**
+ * {@link MapValueOperator} targeting {@link RealmModel}s values in {@link RealmMap}s.
+ */
+class RealmModelValueOperator<T> extends MapValueOperator<T> {
+
+    RealmModelValueOperator(BaseRealm baseRealm, OsMap osMap, ClassContainer classContainer) {
+        super(baseRealm, osMap, classContainer);
+    }
 
     @Override
-    public void remove(Object key) {
-        osMap.remove(key);
+    public T get(Object key) {
+        long realmModelKey = osMap.getModelRowKey(key);
+        if (realmModelKey == -1) {
+            return null;
+        }
+
+        //noinspection unchecked
+        return (T) baseRealm.get((Class<? extends RealmModel>) classContainer.getClazz(), classContainer.getClassName(), realmModelKey);
+    }
+
+    @Override
+    public T put(Object key, @Nullable T value) {
+        //noinspection unchecked
+        Class<T> clazz = (Class<T>) classContainer.getClazz();
+        String className = classContainer.getClassName();
+        long rowModelKey = osMap.getModelRowKey(key);
+
+        RealmModel realmObject = (RealmModel) value;
+        boolean copyObject;
+        if (value == null) {
+            osMap.put(key, null);
+        } else {
+            // TODO: figure out how to do this with Mixed, check Java_io_realm_internal_core_NativeMixed_nativeCreateMixedLink
+            copyObject = checkCanObjectBeCopied(baseRealm, realmObject, classContainer);
+            RealmObjectProxy proxy = (RealmObjectProxy) ((copyObject) ? copyToRealm((RealmModel) value) : realmObject);
+            osMap.putRow(key, proxy.realmGet$proxyState().getRow$realm().getObjectKey());
+        }
+
+        if (rowModelKey == -1) {
+            return null;
+        } else {
+            //noinspection unchecked
+            return (T) baseRealm.get((Class<? extends RealmModel>) clazz, className, rowModelKey);
+        }
+    }
+
+    // TODO: unify this method and the one in RealmModelListOperator
+    private boolean checkCanObjectBeCopied(BaseRealm realm, RealmModel object, ClassContainer classContainer) {
```

Comment:
Do maps allow embedded objects? Shouldn't we check if it is an embedded object?

Reply:
It should not be allowed in phase 1, so I will add a check.